pkg/middlewares/headers/headers.go

```diff
@@ -178,6 +187,24 @@ func (s *Header) modifyRequestHeaders(req *http.Request) {
 	}
 }
 
+func (s *Header) inlineModifyResponseHeaders(rw http.ResponseWriter, req *http.Request) {
+	originHeader := req.Header.Get("Origin")
+	allowOrigin := s.getAllowOrigin(originHeader)
+
+	if allowOrigin != "" {
+		rw.Header().Set("Access-Control-Allow-Origin", allowOrigin)
+	}
+
+	if s.headers.AccessControlAllowCredentials {
+		rw.Header().Set("Access-Control-Allow-Credentials", "true")
+	}
+
+	if len(s.headers.AccessControlExposeHeaders) > 0 {
+		exposeHeaders := strings.Join(s.headers.AccessControlExposeHeaders, ",")
+		rw.Header().Set("Access-Control-Expose-Headers", exposeHeaders)
+	}
+}
+
 // ModifyResponseHeaders set or delete response headers
 func (s *Header) ModifyResponseHeaders(res *http.Response) error {
```

Comment:
One problem I see is that we're now relying on the fact that during the whole lifetime of a request, both *Header.ServeHTTP and *Header.ModifyResponseHeaders are called, which is why we can get away with doing some of the response processing during  *Header.ServeHTTP.
However, I wouldn't be surprised that we shoot ourselves in the foot later on (in another context), where we call ModifyResponseHeaders _only_ and expect it to do all the work (especially given its name and doc).
Maybe that's enough for now to document better ModifyResponseHeaders, and maybe rename it though. WDYT?

Reply:
I agree with you, in that we _may_ run into issues down the road, but I don't think that there is a better solution for the time being.

I have added more documentation in the comment, and renamed the methods for clarity.